File: backend/app/schemas/supply_chain.py
"""
Pydantic schemas for supply chain visualization.
Tracks supply chain nodes and transitions between them.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class SupplyChainNodeType(str, Enum):
    """Type of supply chain node."""
    PRODUCER = 'PRODUCER'
    PROCESSOR = 'PROCESSOR'
    WAREHOUSE = 'WAREHOUSE'
    DISTRIBUTOR = 'DISTRIBUTOR'
    RETAILER = 'RETAILER'
    CONSUMER = 'CONSUMER'


class GeoCoordinates(BaseModel):
    """Geographic coordinates for map display."""
    lat: float = Field(..., ge=-90, le=90, description="Latitude")
    lng: float = Field(..., ge=-180, le=180, description="Longitude")


# ============================================================
# NODE SCHEMAS
# ============================================================

class SupplyChainNodeCreate(BaseModel):
    """Request to create a supply chain node."""
    product_id: Optional[str] = None
    node_type: SupplyChainNodeType
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    location: Optional[str] = Field(None, max_length=500)
    coordinates: Optional[GeoCoordinates] = None
    order_index: Optional[int] = Field(None, ge=0)
    contact_name: Optional[str] = Field(None, max_length=200)
    contact_phone: Optional[str] = Field(None, max_length=50)
    contact_email: Optional[str] = Field(None, max_length=255)
    external_id: Optional[str] = Field(None, max_length=100)
    image_url: Optional[str] = None
    certificate_urls: Optional[List[str]] = Field(default_factory=list)
    metadata: Optional[dict] = None


class SupplyChainNodeUpdate(BaseModel):
    """Request to update a supply chain node."""
    node_type: Optional[SupplyChainNodeType] = None
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    location: Optional[str] = Field(None, max_length=500)
    coordinates: Optional[GeoCoordinates] = None
    order_index: Optional[int] = Field(None, ge=0)
    contact_name: Optional[str] = Field(None, max_length=200)
    contact_phone: Optional[str] = Field(None, max_length=50)
    contact_email: Optional[str] = Field(None, max_length=255)
    external_id: Optional[str] = Field(None, max_length=100)
    image_url: Optional[str] = None
    certificate_urls: Optional[List[str]] = None
    is_verified: Optional[bool] = None
    metadata: Optional[dict] = None


class SupplyChainNode(BaseModel):
    """A node in the supply chain."""
    id: str
    organization_id: str
    product_id: Optional[str] = None
    node_type: SupplyChainNodeType
    name: str
    description: Optional[str] = None
    location: Optional[str] = None
    coordinates: Optional[GeoCoordinates] = None
    order_index: int
    contact_name: Optional[str] = None
    contact_phone: Optional[str] = None
    contact_email: Optional[str] = None
    external_id: Optional[str] = None
    is_verified: bool
    verified_at: Optional[datetime] = None
    image_url: Optional[str] = None
    certificate_urls: List[str] = Field(default_factory=list)
    metadata: Optional[dict] = None
    created_at: datetime
    updated_at: datetime


# ============================================================
# STEP SCHEMAS
# ============================================================

class SupplyChainStepCreate(BaseModel):
    """Request to create a supply chain step (transition between nodes)."""
    from_node_id: str
    to_node_id: str
    description: Optional[str] = Field(None, max_length=2000)
    transport_method: Optional[str] = Field(None, max_length=100)
    distance_km: Optional[float] = Field(None, ge=0)
    duration_hours: Optional[float] = Field(None, ge=0)
    timestamp: Optional[datetime] = None
    expected_arrival: Optional[datetime] = None
    tracking_number: Optional[str] = Field(None, max_length=100)
    batch_id: Optional[str] = Field(None, max_length=100)
    temperature_celsius: Optional[float] = None
    humidity_percent: Optional[float] = Field(None, ge=0, le=100)
    document_urls: Optional[List[str]] = Field(default_factory=list)
    metadata: Optional[dict] = None


class SupplyChainStepUpdate(BaseModel):
    """Request to update a supply chain step."""
    description: Optional[str] = Field(None, max_length=2000)
    transport_method: Optional[str] = Field(None, max_length=100)
    distance_km: Optional[float] = Field(None, ge=0)
    duration_hours: Optional[float] = Field(None, ge=0)
    timestamp: Optional[datetime] = None
    expected_arrival: Optional[datetime] = None
    verified: Optional[bool] = None
    verification_notes: Optional[str] = Field(None, max_length=2000)
    tracking_number: Optional[str] = Field(None, max_length=100)
    batch_id: Optional[str] = Field(None, max_length=100)
    temperature_celsius: Optional[float] = None
    humidity_percent: Optional[float] = Field(None, ge=0, le=100)
    document_urls: Optional[List[str]] = None
    metadata: Optional[dict] = None


class SupplyChainStep(BaseModel):
    """A step/transition in the supply chain."""
    id: str
    product_id: str
    from_node_id: str
    to_node_id: str
    description: Optional[str] = None
    transport_method: Optional[str] = None
    distance_km: Optional[float] = None
    duration_hours: Optional[float] = None
    timestamp: Optional[datetime] = None
    expected_arrival: Optional[datetime] = None
    verified: bool
    verified_at: Optional[datetime] = None
    verification_notes: Optional[str] = None
    tracking_number: Optional[str] = None
    batch_id: Optional[str] = None
    temperature_celsius: Optional[float] = None
    humidity_percent: Optional[float] = None
    document_urls: List[str] = Field(default_factory=list)
    metadata: Optional[dict] = None
    created_at: datetime
    updated_at: datetime


# ============================================================
# JOURNEY SCHEMAS
# ============================================================

class SupplyChainJourneyNode(BaseModel):
    """Node with step information for journey view."""
    node: SupplyChainNode
    step_to_next: Optional[SupplyChainStep] = None


class SupplyChainJourney(BaseModel):
    """Complete supply chain journey for a product."""
    product_id: str
    product_name: str
    product_slug: str
    organization_id: str
    organization_name: str
    nodes: List[SupplyChainJourneyNode] = Field(default_factory=list)
    total_nodes: int = 0
    verified_nodes: int = 0
    total_steps: int = 0
    verified_steps: int = 0
    total_distance_km: float = 0.0
    total_duration_hours: float = 0.0


class SupplyChainStats(BaseModel):
    """Statistics about a supply chain."""
    total_nodes: int = 0
    verified_nodes: int = 0
    total_steps: int = 0
    verified_steps: int = 0
    total_distance_km: float = 0.0
    total_duration_hours: float = 0.0
    verification_percentage: float = 0.0

File: frontend/src/types/supply-chain.ts
// Supply Chain Visualization Types

export type SupplyChainNodeType =
  | 'PRODUCER'
  | 'PROCESSOR'
  | 'WAREHOUSE'
  | 'DISTRIBUTOR'
  | 'RETAILER'
  | 'CONSUMER'

export interface GeoCoordinates {
  lat: number
  lng: number
}

export interface SupplyChainNode {
  id: string
  organization_id: string
  product_id?: string | null
  node_type: SupplyChainNodeType
  name: string
  description?: string | null
  location?: string | null
  coordinates?: GeoCoordinates | null
  order_index: number
  contact_name?: string | null
  contact_phone?: string | null
  contact_email?: string | null
  external_id?: string | null
  is_verified: boolean
  verified_at?: string | null
  image_url?: string | null
  certificate_urls: string[]
  metadata?: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export interface SupplyChainStep {
  id: string
  product_id: string
  from_node_id: string
  to_node_id: string
  description?: string | null
  transport_method?: string | null
  distance_km?: number | null
  duration_hours?: number | null
  timestamp?: string | null
  expected_arrival?: string | null
  verified: boolean
  verified_at?: string | null
  verification_notes?: string | null
  tracking_number?: string | null
  batch_id?: string | null
  temperature_celsius?: number | null
  humidity_percent?: number | null
  document_urls: string[]
  metadata?: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export interface SupplyChainJourneyNode {
  node: SupplyChainNode
  step_to_next?: SupplyChainStep | null
}

export interface SupplyChainJourney {
  product_id: string
  product_name: string
  product_slug: string
  organization_id: string
  organization_name: string
  nodes: SupplyChainJourneyNode[]
  total_nodes: number
  verified_nodes: number
  total_steps: number
  verified_steps: number
  total_distance_km: number
  total_duration_hours: number
}

export interface SupplyChainStats {
  total_nodes: number
  verified_nodes: number
  total_steps: number
  verified_steps: number
  total_distance_km: number
  total_duration_hours: number
  verification_percentage: number
}

// Create/Update DTOs
export interface SupplyChainNodeCreate {
  product_id?: string | null
  node_type: SupplyChainNodeType
  name: string
  description?: string | null
  location?: string | null
  coordinates?: GeoCoordinates | null
  order_index?: number | null
  contact_name?: string | null
  contact_phone?: string | null
  contact_email?: string | null
  external_id?: string | null
  image_url?: string | null
  certificate_urls?: string[]
  metadata?: Record<string, unknown> | null
}

export interface SupplyChainNodeUpdate {
  node_type?: SupplyChainNodeType
  name?: string
  description?: string | null
  location?: string | null
  coordinates?: GeoCoordinates | null
  order_index?: number
  contact_name?: string | null
  contact_phone?: string | null
  contact_email?: string | null
  external_id?: string | null
  image_url?: string | null
  certificate_urls?: string[]
  is_verified?: boolean
  metadata?: Record<string, unknown> | null
}

export interface SupplyChainStepCreate {
  from_node_id: string
  to_node_id: string
  description?: string | null
  transport_method?: string | null
  distance_km?: number | null
  duration_hours?: number | null
  timestamp?: string | null
  expected_arrival?: string | null
  tracking_number?: string | null
  batch_id?: string | null
  temperature_celsius?: number | null
  humidity_percent?: number | null
  document_urls?: string[]
  metadata?: Record<string, unknown> | null
}

export interface SupplyChainStepUpdate {
  description?: string | null
  transport_method?: string | null
  distance_km?: number | null
  duration_hours?: number | null
  timestamp?: string | null
  expected_arrival?: string | null
  verified?: boolean
  verification_notes?: string | null
  tracking_number?: string | null
  batch_id?: string | null
  temperature_celsius?: number | null
  humidity_percent?: number | null
  document_urls?: string[]
  metadata?: Record<string, unknown> | null
}

// Node type configuration for UI
export interface NodeTypeConfig {
  type: SupplyChainNodeType
  label: string
  icon: string
  color: string
  bgColor: string
}

export const NODE_TYPE_CONFIGS: Record<SupplyChainNodeType, NodeTypeConfig> = {
  PRODUCER: {
    type: 'PRODUCER',
    label: 'Производитель',
    icon: 'Leaf',
    color: 'text-green-600',
    bgColor: 'bg-green-100 dark:bg-green-900/30',
  },
  PROCESSOR: {
    type: 'PROCESSOR',
    label: 'Переработка',
    icon: 'Factory',
    color: 'text-blue-600',
    bgColor: 'bg-blue-100 dark:bg-blue-900/30',
  },
  WAREHOUSE: {
    type: 'WAREHOUSE',
    label: 'Склад',
    icon: 'Warehouse',
    color: 'text-amber-600',
    bgColor: 'bg-amber-100 dark:bg-amber-900/30',
  },
  DISTRIBUTOR: {
    type: 'DISTRIBUTOR',
    label: 'Дистрибьютор',
    icon: 'Truck',
    color: 'text-cyan-600',
    bgColor: 'bg-cyan-100 dark:bg-cyan-900/30',
  },
  RETAILER: {
    type: 'RETAILER',
    label: 'Розница',
    icon: 'Store',
    color: 'text-purple-600',
    bgColor: 'bg-purple-100 dark:bg-purple-900/30',
  },
  CONSUMER: {
    type: 'CONSUMER',
    label: 'Потребитель',
    icon: 'User',
    color: 'text-pink-600',
    bgColor: 'bg-pink-100 dark:bg-pink-900/30',
  },
}
